### PR TITLE
[IMP] Unique tag names

### PR DESCRIPTION
**Pull request by atp-odoo**
Task: https://www.odoo.com/web#id=8812&view_type=form&model=project.task&menu_id=3942&action=327
pad : https://pad.odoo.com/p/r.00aDOS0i0SOXSuwv

### CODE REVIEW

**Comment by rim-odoo** [comment]
Code is conflicting, can you rebase it ?

**Comment by msa-odoo** [comment]
I have resolved conflicts and rebase it. thanks

File: openerp/models.py [hunk 1]
```diff
@@ -4844,6 +4846,28 @@ def _check_recursion(self, cr, uid, ids, context=None, parent=None):
...
+        meaning of input_str and work only for some cases"""
+        input_str = ustr(input_str)
+        nkfd_form = unicodedata.normalize('NFKD', input_str)
+        return u''.join([c for c in nkfd_form if not unicodedata.combining(c)])
+
+    # _constraint method making all tag a unique case and accent insensitive
+    @api.multi
+    def _check_unique_accent(self):
+        for tag in self:
+            unaccent = expression.get_unaccent_wrapper(self._cr)
```
Comment (vme-odoo) [inline openerp/models.py]: remove unaccent from looop

File: addons/calendar/calendar.py [hunk 2]
```diff
@@ -680,7 +680,7 @@ class calendar_event_type(osv.Model):
     _columns = {
         'name': fields.char('Name', required=True, translate=True),
     }
-
+    _constraints = [(osv.Model._check_unique_accent, _('Error! Tag name already exist.'), ['name'])]
```
Comment (tde-banana-odoo) [inline addons/calendar/calendar.py]: already existS . Please update all messages.

File: openerp/addons/base/tests/base_test.yml [hunk 3]
```diff
@@ -92,56 +92,6 @@
         name: Child 2-1
         parent_id: test_categ_2
 -
-    2. Duplicate the parent category and verify that the children have been duplicated too and are below the new parent
```
Comment (tde-banana-odoo) [inline openerp/addons/base/tests/base_test.yml]: Why did you remove those tests ?

File: openerp/addons/base/tests/test_base.py [hunk 4]
```diff
@@ -441,26 +441,6 @@ def test_101_create_translated_record(self):
         fr_context_cat = self.res_category.browse(cr, uid, self.cat_id, context={'lang':'fr_FR'})
         self.assertEqual(fr_context_cat.name, 'Clients', "Translation not found")
 
-    def test_102_duplicate_record(self):
```
Comment (tde-banana-odoo) [inline openerp/addons/base/tests/test_base.py]: Why did you remove those tests ?

File: openerp/models.py [hunk 5]
```diff
@@ -48,6 +48,8 @@
 import pytz
 import re
 import time
+import unicodedata
+from openerp.tools import ustr
```
Comment (tde-banana-odoo) [inline openerp/models.py]: Please order your imports.

File: openerp/models.py [hunk 6]
```diff
@@ -4889,6 +4891,28 @@ def _check_recursion(self, cr, uid, ids, context=None, parent=None):
                 if current_id == id:
                     return False
         return True
+    
+    # Inspired by http://stackoverflow.com/questions/517923
+    @api.model
+    def remove_accents(self, input_str):
```
Comment (tde-banana-odoo) [inline openerp/models.py]: Why defining a method on model ? It should probably be in tools.

File: openerp/models.py [hunk 7]
```diff
@@ -4889,6 +4891,28 @@ def _check_recursion(self, cr, uid, ids, context=None, parent=None):
...
+        input_str = ustr(input_str)
+        nkfd_form = unicodedata.normalize('NFKD', input_str)
+        return u''.join([c for c in nkfd_form if not unicodedata.combining(c)])
+
+    # _constraint method making all tag a unique case and accent insensitive
+    @api.multi
+    def _check_unique_accent(self):
+        unaccent = expression.get_unaccent_wrapper(self._cr)
+        for tag in self:
+            query = ("""SELECT id from {table_name} WHERE lower({name}) = lower({tag}) and id != {current_id}""".format(table_name=str(self._model).replace('.','_'), name=unaccent('name'), tag=unaccent('%s'), current_id=tag.id))
```
Comment (tde-banana-odoo) [inline openerp/models.py]: self._model.replace does not seem valid to me; if the table is specified and different from model, it will crash. Isn't it a _table variable like in the old API ?

Comment (tde-banana-odoo) [inline openerp/models.py]: I think you can also add a limit on the select, one is sufficient to make the test False.

File: openerp/models.py [hunk 8]
```diff
@@ -4889,6 +4891,28 @@ def _check_recursion(self, cr, uid, ids, context=None, parent=None):
...
+        """Suboptimal-but-better-than-nothing way to replace accented
+        latin letters by an ASCII equivalent. Will obviously change the
+        meaning of input_str and work only for some cases"""
+        input_str = ustr(input_str)
+        nkfd_form = unicodedata.normalize('NFKD', input_str)
+        return u''.join([c for c in nkfd_form if not unicodedata.combining(c)])
+
+    # _constraint method making all tag a unique case and accent insensitive
+    @api.multi
+    def _check_unique_accent(self):
```
Comment (tde-banana-odoo) [inline openerp/models.py]: I would like some tests for this new constraint. Please add them !

File: openerp/addons/base/tests/base_test.yml (hunk 3 above)
Comment (msa-odoo) [inline openerp/addons/base/tests/base_test.yml]: Here we making category as a unique and this test case is for duplicate the parent category and verify that the children have been duplicated too and are below the new parent.
So now we are not allowed to create duplicate category so i remove this tests as no more needed.

File: openerp/addons/base/tests/test_base.py (hunk 4 above)
Comment (msa-odoo) [inline openerp/addons/base/tests/test_base.py]: Same as above here also we making copy of the category and checks that duplication did set untranslated value or not.
So now we are not allowed to create duplicate category so i remove this tests as no more needed.

File: openerp/models.py [hunk 9]
```diff
@@ -4890,6 +4890,18 @@ def _check_recursion(self, cr, uid, ids, context=None, parent=None):
...
+        unaccent = expression.get_unaccent_wrapper(self._cr)
+        for tag in self:
+            query = ("""SELECT id from {table_name} WHERE lower({name}) = lower({tag}) and id != {current_id} LIMIT 1""".format(table_name=str(self._table), name=unaccent('name'), tag=unaccent('%s'), current_id=tag.id))
+            self._cr.execute(query,[tag.name])
+            tag_ids = self._cr.fetchall()
+            if tag_ids:
+                return False
+        return True
```
Comment (vme-odoo) [inline openerp/models.py]: code can be lesser here.

File: openerp/addons/base/tests/test_tag.py [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+        # Creating same category name should not allowed 
+        self.assertRaises(ValidationError, ResPartnerCategory.create, {'name':'Test'})
+        # Creating same category name (case insencitive) should not allowed
+        self.assertRaises(ValidationError, ResPartnerCategory.create, {'name':'test'})
+        has_unaccent = openerp.modules.db.has_unaccent(self.env.cr)
+        # if unaccent is installed in postgress and option --unaccent is given than accented same category name should not allowed 
+        if openerp.tools.config['unaccent'] and has_unaccent:
+            self.assertRaises(ValidationError, ResPartnerCategory.create, {'name':'Tést'})
+        else:
+            ResPartnerCategory.create({'name':'Tést'})
```
Comment (vme-odoo) [inline openerp/addons/base/tests/test_tag.py]: improve comments

File: openerp/models.py (hunk 9 above)
Comment (vme-odoo) [inline openerp/models.py]: use api one.

File: addons/website_forum/models/forum.py [hunk 11]
```diff
@@ -149,21 +150,27 @@ def create(self, values):
...
-                    if user.exists() and user.karma >= self.karma_retag:
+                self._cr.execute(query,[tag[1:]])
+                existing_tag_ids = self._cr.fetchall()
+                if existing_tag_ids and existing_tag_ids[0][0] not in existing_keep:
+                    existing_keep.append(existing_tag_ids[0][0])
+                # check if not found in existing tag and user have Karma needed to create need tag
+                elif not existing_tag_ids and user.exists() and user.karma >= self.karma_retag:
+                    if post_tags:
+                        [post_tags.append((0, 0, {'name': tag[1:], 'forum_id': self.id})) for post_tag in post_tags if not tools.remove_accents(post_tag[2]['name']).lower() == tools.remove_accents(tag[1:]).lower()]
+                    else:
```
Comment (rim-odoo) [inline addons/website_forum/models/forum.py]: @tde-banana-odoo Can I have your opinion about that?

File: openerp/models.py [hunk 12]
```diff
@@ -4890,6 +4890,24 @@ def _check_recursion(self, cr, uid, ids, context=None, parent=None):
...
+            return rec._check_unique_accent()
+        except ValidationError:
+            return False
+    @api.one
+    @api.v8
+    def _check_unique_accent(self):
+        unaccent = expression.get_unaccent_wrapper(self._cr)
+        query = ("""SELECT id from {table_name} WHERE lower({name}) = lower({tag}) and id != {current_id} LIMIT 1""".format(table_name=str(self._table), name=unaccent('name'), tag=unaccent('%s'), current_id=('%s')))
+        self._cr.execute(query, [self.name, self.id])
+        if self._cr.fetchall():
```
Comment (rim-odoo) [inline openerp/models.py]: do you need to fetch the results? can't you simply count and check if there are results? it could be faster

File: addons/mail/mail_alias.py [hunk 13]
```diff
@@ -23,6 +23,7 @@
 import re
 import unicodedata
```
Comment (rim-odoo) [inline addons/mail/mail_alias.py]: I think this import is unuseful

File: openerp/models.py (hunk 12 above)
Comment (msa-odoo) [inline openerp/models.py]: Yes it could be faster, if we use search_count() but here we used unaccent function of PostgreSQL on column "name", so we can not able to apply this type of domain in search_count() that is why here we execute query and fetch result.

And about self._cr.fetchall() can we use self._cr.rowcount ? Rowcount will return the number of rows in the query.

Comment (rim-odoo) [inline openerp/models.py]: Ok, thanks for the explanation

File: addons/website_forum/models/forum.py [hunk 14]
```diff
@@ -149,21 +150,27 @@ def create(self, values):
     def _tag_to_write_vals(self, tags=''):
         User = self.env['res.users']
         Tag = self.env['forum.tag']
+        user = User.sudo().browse(self._uid)
```
Comment (tde-banana-odoo) [inline addons/website_forum/models/forum.py]: Why this ? All users can read themselves normally (at least some fields).

File: openerp/addons/base/tests/base_test.yml (hunk 3 above)
Comment (tde-banana-odoo) [inline openerp/addons/base/tests/base_test.yml]: I still don't understand why this test should be removed. You should be able to duplicate a category; maybe the name has to be modified when copying.

File: openerp/addons/base/tests/test_base.py (hunk 4 above)
Comment (tde-banana-odoo) [inline openerp/addons/base/tests/test_base.py]: I still don't understand why this test should be removed. You should be able to duplicate a category; maybe the name has to be modified when copying.

File: openerp/tools/mail.py [hunk 15]
```diff
@@ -596,6 +597,14 @@ def append_content_to_html(html, content, plaintext=True, preserve=False, contai
         return '%s%s' % (html, content)
     return '%s%s%s' % (html[:insert_location], content, html[insert_location:])
 
+# Inspired by http://stackoverflow.com/questions/517923
+def remove_accents(input_str):
```
Comment (tde-banana-odoo) [inline openerp/tools/mail.py]: Why in mail ?

Comment (msa-odoo) [inline openerp/tools/mail.py]: In tools/mail.py Because we are make this function generic and used in mail  as well as in website_forum.
